src/loose_deserializer.rs

```diff
@@ -0,0 +1,203 @@
+use std::convert::TryInto;
+use std::io::Read;
+
+use crate::de::{Enum, MapAccess, SeqAccess};
+use crate::error::{Error, Result};
+use crate::value::layout::Summand;
+use crate::value::layout::{BinProtRule, BranchingIterator};
+use crate::Deserializer as DS;
+use crate::ReadBinProtExt;
+use serde::de::{Deserializer, Visitor};
+use serde::{Deserialize, Serialize};
+
+use byteorder::ReadBytesExt;
+
+impl<'de, 'a, R: Read> DS<R> {
+    pub fn deserialize_loose<V>(&mut self, visitor: V) -> Result<V::Value>
+    where
+        V: Visitor<'de>,
+    {
+        if let Some(iter) = &mut self.layout_iter {
+            loop {
+                match iter.next() {
+                    Ok(Some(rule)) => {
+                        match rule {
+                            BinProtRule::Unit => {
+                                return self.deserialize_unit(visitor);
+                            }
+                            BinProtRule::Record(fields) => {
+                                // Grab the field names from the rule to pass to the map access
+                                return visitor.visit_map(MapAccess::new(
+                                    self,
+                                    fields.into_iter().map(|f| f.field_name).rev().collect(),
+                                ));
+                            }
+                            BinProtRule::Tuple(items) => {
+                                return visitor.visit_seq(SeqAccess::new(self, items.len()));
+                            }
+                            BinProtRule::Sum(summands) => {
+                                // read the enum variant index. We need it to
+                                let index = self.rdr.bin_read_variant_index()?;
+                                iter.branch(index.into()).expect("invalid branch index");
+                                return visitor.visit_enum(ValueEnum::new(
+                                    self,
+                                    summands[index as usize].clone(),
+                                ));
+                            }
+                            BinProtRule::Bool => {
+                                return self.deserialize_bool(visitor);
+                            }
+                            BinProtRule::Option(_) => {
+                                return self.deserialize_option(visitor);
+                            }
+                            BinProtRule::Reference(_) => {} // next
```

Comment:
is that a TODO?

Reply:
Nah it literally just does nothing and goes to the next item in the iterator